### DISCUSSION TITLE
Interfaces can have private methods

### DISCUSSION THREAD

**Discussion by giscus** [linkdotnet/Blog.Discussions#88] [General]
# Interfaces can have private methods

Let's drop some "useless" knowledge here. Interfaces can have private methods. This comes with the C# 8 feature: "Default interface methods".

https://steven-giesel.com/blogPost/03957748-73b3-40f9-a9f7-cc95501a3a8e/interfaces-can-have-private-methods

**Comment by nagababuthota984** [comment]
The inclusion of method definitions in an interface also seems to bring the definition and use case of an interface closer to that of an abstract class.

**Reply by toupswork** [reply]
I remember when Microsoft introduced this feature. They also released default method parameters at the same time to appease those devs who had to write a ton of code to do COM interop with MS Office. Internally, there was a lot of disagreement with these choices, with many cringing at the idea

**Reply by linkdotnet** [reply]
I do see the appeal in such Features - even when I personally dislike it (even as library maintainer)
Abstract interfaces just wash out the lines between interface and abstract classes.

**Reply by toupswork** [reply]
I guess they acquiesced to developer demand over purism. But I think it's mostly because they didn't have a way to allow you to alter your interface without causing consumers of that interface to recompile.
